src/common/src/array/list_array.rs

```diff
@@ -624,6 +626,197 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
+    /// Construct an array from literal string.
+    pub fn from_str(input: &str, data_type: &DataType) -> Result<Self, String> {
+        struct Parser<'a> {
+            input: &'a str,
+            data_type: &'a DataType,
+        }
+
+        impl Parser<'_> {
+            /// Parse a datum.
+            fn parse(&mut self) -> Result<Datum, String> {
+                self.skip_whitespace();
+                if self.try_parse_null() {
+                    return Ok(None);
+                }
+                if self.data_type.is_array() {
+                    Ok(Some(self.parse_array()?.into()))
+                } else {
+                    self.parse_value()
+                }
+            }
+
+            /// Parse an array.
+            fn parse_array(&mut self) -> Result<ListValue, String> {
+                self.skip_whitespace();
+                if !self.try_consume('{') {
+                    return Err("Array value must start with \"{\"".to_string());
+                }
+                self.skip_whitespace();
+                if self.try_consume('}') {
+                    return Ok(ListValue::new(vec![]));
+                }
+                let mut elems = Vec::new();
+                loop {
+                    let mut parser = Self {
+                        input: self.input,
+                        data_type: self.data_type.as_list(),
+                    };
+                    elems.push(parser.parse()?);
+                    self.input = parser.input;
+
+                    // expect ',' or '}'
+                    self.skip_whitespace();
+                    match self.peek() {
+                        Some(',') => {
+                            self.try_consume(',');
+                        }
+                        Some('}') => {
+                            self.try_consume('}');
+                            break;
+                        }
+                        None => return Err(Self::eoi()),
+                        _ => return Err("Unexpected array element.".to_string()),
+                    }
+                }
+                Ok(ListValue::new(elems))
+            }
+
+            /// Parse a scalar value.
+            fn parse_value(&mut self) -> Result<Datum, String> {
+                if self.peek() == Some('"') {
+                    return self.parse_quoted();
+                }
+                // peek until the next unescaped ',' or '}'
+                let mut chars = self.input.char_indices();
+                let mut has_escape = false;
+                let s = loop {
+                    match chars.next().ok_or_else(Self::eoi)? {
+                        (_, '\\') => {
+                            has_escape = true;
+                            chars.next().ok_or_else(Self::eoi)?;
+                        }
+                        (i, c @ ',' | c @ '}') => {
+                            let s = self.input[..i].trim_end();
+                            self.input = &self.input[i..];
```

Comment:
Would like a comment here that `self.input` is to inform parent `parse_array` about the continuation location.